File: SpearoGo/Services/SolunarService.swift
import Foundation
import CoreLocation

// Solunar calculator — pure math, no API required, global coverage.
// Based on John Alden Knight's solunar theory:
//   Major periods: moon transit + anti-transit (~2h each)
//   Minor periods: moonrise + moonset (~1h each)
// Moon position computed via simplified orbital elements (Meeus, Ch.48).

struct SolunarService {
    func calculate(coordinate: CLLocationCoordinate2D, date: Date = Date()) -> SolunarData {
        let jd = julianDay(from: date)
        let moonPos = moonPosition(jd: jd)
        let sunPos  = sunPosition(jd: jd)

        let moonIllum = illumination(moonLon: moonPos.longitude, sunLon: sunPos.longitude)
        let moonPhase = moonPhaseValue(jd: jd)

        let (moonrise, moonset) = riseSet(
            declination: moonPos.declination,
            ra: moonPos.rightAscension,
            latitude: coordinate.latitude,
            longitude: coordinate.longitude,
            jd: jd,
            isMoon: true
        )

        let (sunrise, sunset) = riseSet(
            declination: sunPos.declination,
            ra: sunPos.rightAscension,
            latitude: coordinate.latitude,
            longitude: coordinate.longitude,
            jd: jd,
            isMoon: false
        )

        // Major periods: ±1h around moon transit (upper/lower culmination)
        let transit = moonTransit(ra: moonPos.rightAscension, longitude: coordinate.longitude, jd: jd)
        let antiTransit = transit.addingTimeInterval(6 * 3600)

        // Minor periods: around moonrise/moonset
        let nextMajor: Date? = [transit, antiTransit]
            .filter { $0 > date.addingTimeInterval(-3600) }
            .min()

        let nextMinor: Date? = [moonrise, moonset]
            .compactMap { $0 }
            .filter { $0 > date.addingTimeInterval(-1800) }
            .min()

        let rating = activityRating(phase: moonPhase, illum: moonIllum, nextMajor: nextMajor, date: date)

        return SolunarData(
            moonPhase:        moonPhase,
            moonIllumination: moonIllum,
            moonrise:         moonrise,
            moonset:          moonset,
            sunrise:          sunrise,
            sunset:           sunset,
            nextMajorPeriod:  nextMajor,
            nextMinorPeriod:  nextMinor,
            activityRating:   rating,
            fetchedAt:        Date()
        )
    }

    // MARK: - Julian Day

    private func julianDay(from date: Date) -> Double {
        date.timeIntervalSince1970 / 86400.0 + 2440587.5
    }

    // MARK: - Moon position (low-precision, Meeus simplified)

    private struct CelestialPosition {
        let longitude: Double      // degrees
        let rightAscension: Double // hours
        let declination: Double    // degrees
    }

    private func moonPosition(jd: Double) -> CelestialPosition {
        let T = (jd - 2451545.0) / 36525.0
        let L0 = (218.316 + 13.176396 * (jd - 2451545.0)).truncatingRemainder(dividingBy: 360)
        let M  = (134.963 + 13.064993 * (jd - 2451545.0)).truncatingRemainder(dividingBy: 360)
        let F  = (93.272  + 13.229350 * (jd - 2451545.0)).truncatingRemainder(dividingBy: 360)

        let lon = L0 + 6.289 * sin(M * .pi / 180)
                     - 1.274 * sin((2 * F - M) * .pi / 180)
                     + 0.658 * sin(2 * F * .pi / 180)
                     - 0.214 * sin(2 * M * .pi / 180)
                     - 0.186 * sin((M - 0.0003 * T) * .pi / 180)

        let lat = 5.128 * sin(F * .pi / 180)

        // Equatorial coords
        let e = 23.439 - 0.0000004 * (jd - 2451545.0)
        let ra = atan2(
            sin(lon * .pi / 180) * cos(e * .pi / 180) - tan(lat * .pi / 180) * sin(e * .pi / 180),
            cos(lon * .pi / 180)
        ) * 180 / .pi / 15  // hours

        let dec = asin(
            sin(lat * .pi / 180) * cos(e * .pi / 180)
            + cos(lat * .pi / 180) * sin(e * .pi / 180) * sin(lon * .pi / 180)
        ) * 180 / .pi

        return CelestialPosition(longitude: lon.truncatingRemainder(dividingBy: 360),
                                  rightAscension: (ra + 24).truncatingRemainder(dividingBy: 24),
                                  declination: dec)
    }

    private func sunPosition(jd: Double) -> CelestialPosition {
        let D  = jd - 2451545.0
        let g  = (357.529 + 0.98560028 * D).truncatingRemainder(dividingBy: 360)
        let q  = (280.459 + 0.98564736 * D).truncatingRemainder(dividingBy: 360)
        let lon = q + 1.915 * sin(g * .pi / 180) + 0.020 * sin(2 * g * .pi / 180)
        let e  = 23.439 - 0.00000036 * D
        let ra = atan2(cos(e * .pi / 180) * sin(lon * .pi / 180), cos(lon * .pi / 180)) * 180 / .pi / 15
        let dec = asin(sin(e * .pi / 180) * sin(lon * .pi / 180)) * 180 / .pi
        return CelestialPosition(longitude: lon, rightAscension: (ra + 24).truncatingRemainder(dividingBy: 24), declination: dec)
    }

    // MARK: - Rise / Set times (approximate)

    private func riseSet(declination: Double, ra: Double, latitude: Double,
                         longitude: Double, jd: Double, isMoon: Bool) -> (rise: Date?, set: Date?) {
        let latR = latitude * .pi / 180
        let decR = declination * .pi / 180
        let h0 = isMoon ? -0.833 : -0.833  // standard refraction
        let cosH = (sin(h0 * .pi / 180) - sin(latR) * sin(decR)) / (cos(latR) * cos(decR))
        guard abs(cosH) <= 1 else { return (nil, nil) }  // circumpolar / never rises

        let H = acos(cosH) * 180 / .pi
        let noon = (ra - longitude / 15).truncatingRemainder(dividingBy: 24)
        let riseH = noon - H / 15
        let setH  = noon + H / 15

        func toDate(_ hours: Double) -> Date {
            let midnight = (jd - 0.5).rounded(.down) - 2440587.5  // unix midnight
            return Date(timeIntervalSince1970: midnight * 86400 + hours * 3600)
        }

        return (toDate(riseH), toDate(setH))
    }

    // MARK: - Moon transit

    private func moonTransit(ra: Double, longitude: Double, jd: Double) -> Date {
        let D = jd - 2451545.0
        let GMST = (280.46061837 + 360.98564736629 * D).truncatingRemainder(dividingBy: 360)
        let LST = (GMST + longitude).truncatingRemainder(dividingBy: 360)
        var hourAngle = ra * 15 - LST
        if hourAngle < 0 { hourAngle += 360 }
        let hoursUntilTransit = (360 - hourAngle) / 15.041  // sidereal rate
        return Date().addingTimeInterval(hoursUntilTransit * 3600)
    }

    // MARK: - Illumination

    private func illumination(moonLon: Double, sunLon: Double) -> Double {
        let angle = abs(moonLon - sunLon).truncatingRemainder(dividingBy: 360)
        return (1 - cos(angle * .pi / 180)) / 2
    }

    private func moonPhaseValue(jd: Double) -> Double {
        let D = jd - 2451545.0
        let raw = (D / 29.53058868).truncatingRemainder(dividingBy: 1)
        return raw < 0 ? raw + 1 : raw
    }

    // MARK: - Activity rating

    private func activityRating(phase: Double, illum: Double, nextMajor: Date?, date: Date) -> String {
        // Score 0–1 based on lunar phase (near full/new = best) + proximity to major period
        let phaseScore = cos(2 * .pi * phase) * 0.5 + 0.5  // 1 at new/full, 0 at quarters
        var timeScore: Double = 0.5
        if let major = nextMajor {
            let mins = abs(major.timeIntervalSince(date)) / 60
            timeScore = mins < 60 ? 1.0 : mins < 180 ? 0.7 : 0.4
        }
        let total = phaseScore * 0.5 + timeScore * 0.5
        switch total {
        case 0.75...: return "Excellent"
        case 0.55...: return "Good"
        case 0.35...: return "Fair"
        default:      return "Poor"
        }
    }
}

File: wear/app/src/main/java/com/spearotracker/spearogo/services/SolunarService.kt
package com.spearotracker.spearogo.services

import com.spearotracker.spearogo.models.SolunarData
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.*
import java.util.Calendar

@Singleton
class SolunarService @Inject constructor() {

    private data class CelestialPosition(
        val longitude: Double,       // degrees
        val rightAscension: Double,  // hours
        val declination: Double      // degrees
    )

    fun calculate(latitude: Double, longitude: Double, timeMillis: Long = System.currentTimeMillis()): SolunarData {
        val jd = julianDay(timeMillis)
        val moonPos = moonPosition(jd)
        val sunPos = sunPosition(jd)

        val moonIllum = illumination(moonPos.longitude, sunPos.longitude)
        val moonPhase = moonPhaseValue(jd)

        val (moonrise, moonset) = riseSet(latitude, longitude, timeMillis, isMoon = true)
        val (sunrise, sunset) = riseSet(latitude, longitude, timeMillis, isMoon = false)

        // Major periods: ~2h windows around the Moon's culminations.
        val now = timeMillis

        // Culminations alternate every 12h25m: the Moon crosses the meridian
        // (upper) and the anti-meridian (lower). Both are solved directly
        // rather than derived by adding a fixed offset to the other, because
        // the Moon's motion is not uniform.
        val culminations = listOf(
            nextCulmination(longitude, now, targetHourAngleDeg = 0.0),
            nextCulmination(longitude, now, targetHourAngleDeg = 180.0)
        )

        // Next major period
        val nextMajor = culminations
            .filter { it > now - 3600_000 }
            .minOrNull()

        // Next minor period
        val nextMinor = listOfNotNull(moonrise, moonset)
            .filter { it > now - 1800_000 }
            .minOrNull()

        val rating = activityRating(moonPhase, moonIllum, nextMajor, now)

        return SolunarData(
            moonPhase = moonPhase,
            moonIllumination = moonIllum,
            moonrise = moonrise,
            moonset = moonset,
            sunrise = sunrise,
            sunset = sunset,
            nextMajorPeriod = nextMajor,
            nextMinorPeriod = nextMinor,
            activityRating = rating
        )
    }

    private fun julianDay(timeMillis: Long): Double {
        return timeMillis / 86400000.0 + 2440587.5
    }

    private fun moonPosition(jd: Double): CelestialPosition {
        val d = jd - 2451545.0
        val l0 = (218.316 + 13.176396 * d) % 360
        val m = (134.963 + 13.064993 * d) % 360
        val f = (93.272 + 13.229350 * d) % 360

        val lon = l0 + 6.289 * sin(Math.toRadians(m)) -
                  1.274 * sin(Math.toRadians(2 * f - m)) +
                  0.658 * sin(Math.toRadians(2 * f)) -
                  0.214 * sin(Math.toRadians(2 * m)) -
                  0.186 * sin(Math.toRadians(m - 0.0003 * (jd - 2451545.0) / 36525.0))

        val lat = 5.128 * sin(Math.toRadians(f))

        val e = 23.439 - 0.0000004 * d
        val ra = atan2(
            sin(Math.toRadians(lon)) * cos(Math.toRadians(e)) - tan(Math.toRadians(lat)) * sin(Math.toRadians(e)),
            cos(Math.toRadians(lon))
        ).let { Math.toDegrees(it) / 15.0 }

        val dec = asin(
            sin(Math.toRadians(lat)) * cos(Math.toRadians(e)) +
            cos(Math.toRadians(lat)) * sin(Math.toRadians(e)) * sin(Math.toRadians(lon))
        ).let { Math.toDegrees(it) }

        return CelestialPosition(
            longitude = lon % 360,
            rightAscension = (ra + 24) % 24,
            declination = dec
        )
    }

    private fun sunPosition(jd: Double): CelestialPosition {
        val d = jd - 2451545.0
        val g = (357.529 + 0.98560028 * d) % 360
        val q = (280.459 + 0.98564736 * d) % 360
        val lon = q + 1.915 * sin(Math.toRadians(g)) + 0.020 * sin(Math.toRadians(2 * g))
        val e = 23.439 - 0.00000036 * d
        val ra = atan2(
            cos(Math.toRadians(e)) * sin(Math.toRadians(lon)),
            cos(Math.toRadians(lon))
        ).let { Math.toDegrees(it) / 15.0 }
        val dec = asin(sin(Math.toRadians(e)) * sin(Math.toRadians(lon))).let { Math.toDegrees(it) }
        return CelestialPosition(
            longitude = lon,
            rightAscension = (ra + 24) % 24,
            declination = dec
        )
    }

    /**
     * Altitude of a body above the horizon, in degrees.
     */
    private fun altitude(latitude: Double, longitude: Double, millis: Long, isMoon: Boolean): Double {
        val j = julianDay(millis)
        val pos = if (isMoon) moonPosition(j) else sunPosition(j)
        val d = j - 2451545.0
        val gmst = (280.46061837 + 360.98564736629 * d) % 360
        val hourAngle = (gmst + longitude - pos.rightAscension * 15) % 360
        val latR = Math.toRadians(latitude)
        val decR = Math.toRadians(pos.declination)
        return Math.toDegrees(
            asin(sin(latR) * sin(decR) + cos(latR) * cos(decR) * cos(Math.toRadians(hourAngle)))
        )
    }

    /**
     * Rise and set for the local day containing [timeMillis], found by scanning
     * the body's altitude for crossings of the horizon.
     *
     * The previous implementation solved for an hour angle and then placed it
     * with `ra - longitude / 15`, which never referenced sidereal time and so
     * was not anchored to the date at all. On 2026-08-31 it put sunrise at
     * 17:24 and sunset at 06:54, inverted and thirteen hours out. A scan costs
     * a few hundred cheap evaluations and cannot get the day wrong.
     *
     * Returns nulls when the body does not cross the horizon that day, which is
     * the honest answer inside a polar summer or winter — and for the Moon,
     * simply on days when it does not rise.
     */
    private fun riseSet(
        latitude: Double,
        longitude: Double,
        timeMillis: Long,
        isMoon: Boolean
    ): Pair<Long?, Long?> {
        // Standard refraction for the Sun's upper limb; for the Moon, parallax
        // very nearly cancels semidiameter and refraction.
        val horizon = if (isMoon) 0.125 else -0.833

        val cal = Calendar.getInstance()
        cal.timeInMillis = timeMillis
        cal.set(Calendar.HOUR_OF_DAY, 0)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        val dayStart = cal.timeInMillis

        val step = 5 * 60 * 1000L
        val dayEnd = dayStart + 24 * 3600 * 1000L

        var rise: Long? = null
        var set: Long? = null
        var previousAlt = altitude(latitude, longitude, dayStart, isMoon)
        var t = dayStart + step

        while (t <= dayEnd) {
            val alt = altitude(latitude, longitude, t, isMoon)
            // Linear interpolation across the step gets within a few seconds.
            if (previousAlt < horizon && alt >= horizon && rise == null) {
                rise = t - step + ((horizon - previousAlt) / (alt - previousAlt) * step).toLong()
            }
            if (previousAlt > horizon && alt <= horizon && set == null) {
                set = t - step + ((previousAlt - horizon) / (previousAlt - alt) * step).toLong()
            }
            previousAlt = alt
            t += step
        }

        return Pair(rise, set)
    }

    /**
     * Time of the Moon's next crossing of a given hour angle — 0 for upper
     * culmination, 180 for lower. These are the solunar "major" periods.
     *
     * Solved iteratively. A single pass using the Moon's position *now* is what
     * this code used to do, and it is wrong by up to an hour for a culmination
     * a day away, because the Moon moves about half a degree an hour against
     * the stars. Re-evaluating at the estimate converges in a couple of passes.
     */
    private fun nextCulmination(longitude: Double, fromMillis: Long, targetHourAngleDeg: Double): Long {
        var t = fromMillis
        repeat(4) {
            val j = julianDay(t)
            val ra = moonPosition(j).rightAscension
            val d = j - 2451545.0
            val gmst = (280.46061837 + 360.98564736629 * d) % 360
            val lst = (gmst + longitude) % 360
            // How far past the target the Moon already is, in degrees,
            // normalised to (-180, 180].
            var past = lst - ra * 15 - targetHourAngleDeg
            past = ((past + 180.0) % 360.0 + 360.0) % 360.0 - 180.0
            t += ((-past / SIDEREAL_DEGREES_PER_HOUR) * 3600 * 1000).toLong()
        }
        // Converges on the NEAREST crossing, which may be just behind us.
        while (t < fromMillis - 3600_000) t += LUNAR_DAY_MS
        return t
    }

    private fun illumination(moonLon: Double, sunLon: Double): Double {
        val angle = abs(moonLon - sunLon) % 360
        return (1 - cos(Math.toRadians(angle))) / 2
    }

    private fun moonPhaseValue(jd: Double): Double {
        val d = jd - 2451545.0
        val raw = (d / 29.53058868) % 1
        return if (raw < 0) raw + 1 else raw
    }

    private fun activityRating(phase: Double, illum: Double, nextMajor: Long?, nowMillis: Long): String {
        val phaseScore = cos(2 * PI * phase) * 0.5 + 0.5
        var timeScore = 0.5
        if (nextMajor != null) {
            val mins = abs(nextMajor - nowMillis) / 60000.0
            timeScore = when {
                mins < 60 -> 1.0
                mins < 180 -> 0.7
                else -> 0.4
            }
        }
        val total = phaseScore * 0.5 + timeScore * 0.5
        return when {
            total >= 0.75 -> "Excellent"
            total >= 0.55 -> "Good"
            total >= 0.35 -> "Fair"
            else -> "Poor"
        }
    }
}

// Rate at which the sky turns relative to the Moon, degrees per hour: the Moon
// moves eastward against the stars, so it returns to the meridian every ~24h50m
// rather than every sidereal day.
private const val SIDEREAL_DEGREES_PER_HOUR = 15.041

// A lunar day: the interval between successive upper culminations, 24h50m28s.
private const val LUNAR_DAY_MS = 89_428_320L
